projects/plugins/jetpack/_inc/client/recommendations/product-card-upsell/index.jsx

```diff
@@ -1,56 +1,81 @@
-import { __, sprintf } from '@wordpress/i18n';
+import React, { useEffect, useMemo } from 'react';
+import { connect } from 'react-redux';
 import { getCurrencyObject } from '@automattic/format-currency';
+import { __, sprintf } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
 import Button from 'components/button';
-import analytics from 'lib/analytics';
+import { getIntroOffers, isFetchingIntroOffers } from 'state/intro-offers';
+import { isFetchingSiteDiscount, getSiteDiscount } from 'state/site/reducer';
+import DiscountBadge from '../discount-badge';
+import withUpgradeUrl from '../hoc/with-upgrade-url';
 import RecommendedHeader from '../sidebar/recommended-header';
 
 /**
  * Style dependencies
  */
 import './style.scss';
-import classNames from 'classnames';
 
-const ProductCardUpsell = ( {
+const Price = ( { className, integer, fraction, symbol } ) => (
+	<div className={ className }>
+		<sup className="jp-recommendations-product-card-upsell__currency-symbol">{ symbol }</sup>
+		<span className="jp-recommendations-product-card-upsell__price-integer">{ integer }</span>
+		<sup className="jp-recommendations-product-card-upsell__price-fraction">{ fraction }</sup>
+	</div>
+);
+
+const ProductCardUpsellComponent = ( {
+	slug,
+	title,
+	description,
 	billing_timeframe,
 	cost_timeframe,
-	currency_code,
-	description,
-	features,
-	price,
+	currency_code: currency,
 	isRecommended,
-	product_slug,
-	title,
-	upgradeUrl,
+	features,
 	onClick,
+	onMount,
+	isFetchingDiscount,
+	isFetchingOffers,
+	discountData,
+	introOffers,
+	upgradeUrl,
 } ) => {
-	useEffect( () => {
-		analytics.tracks.recordEvent( 'jetpack_recommendations_summary_sidebar_display', {
-			type: 'upsell_with_price',
-			product_slug,
-		} );
-	}, [ product_slug ] );
-
-	const onUpsellClick = useCallback( () => {
-		analytics.tracks.recordEvent( 'jetpack_recommendations_summary_sidebar_click', {
-			type: 'upsell_with_price',
-			product_slug,
-		} );
-	}, [ product_slug ] );
+	const { discount } = discountData;
+	const { original_price: originalPrice, raw_price: introPrice } = useMemo(
+		() => introOffers.find( ( { product_slug } ) => product_slug === slug ) || {},
+		[ slug, introOffers ]
+	);
+	const finalPrice = discount ? introPrice * ( 1 - discount / 100 ) : introPrice;
+	const totalDiscount = originalPrice
+		? Math.round( ( ( originalPrice - finalPrice ) / originalPrice ) * 100 )
+		: null;
+	const originalCurrencyObject = useMemo( () => getCurrencyObject( originalPrice / 12, currency ), [
+		originalPrice,
+		currency,
+	] );
+	const currencyObject = useMemo( () => getCurrencyObject( finalPrice / 12, currency ), [
+		finalPrice,
+		currency,
+	] );
 
-	const currencyObject = getCurrencyObject( price, currency_code );
 	const header = isRecommended && (
 		<RecommendedHeader className="jp-recommendations-product-card-upsell__header" />
 	);
 
+	useEffect( () => {
+		if ( onMount ) {
+			onMount();
+		}
+	}, [] );
+
```

Comment:
> Something about having an onMount prop feels odd
Just curious, do you know why it feels like this to you?

I've gone through several rounds of refactoring there. Taking a look back at it now, it seems that the events can be tracked directly in the cards ancestors indeed 👍

Reply:
I think it feels like that because it seems like if there is some logic that should be executed when this particular component is mounted, it should live in the component itself instead of being passed in from somewhere else. IMO it makes the code a bit harder to follow since the reader needs to be aware of the context where this component is used to know what happens when it mounts. 

I don't think that's a rule though, just my preference.